### ISSUE TITLE
SetFieldSettings and Recipients Dropdown result in invalid email address being selected

### ISSUE THREAD

**Issue by jimirobaer** [verbb/formie#578]
**Description**

I'm populating a Recipients dropdown with some values through Twig. The values I supply render fine on the front-end. This code is executed before rendering of the form.

```
{% set recipientsArray = [] %}

{% for entry in entries %}
	{% set recipientsField = recipientsArray|merge(
		[{"label": entry.title, "value": entry.email, "isDefault": loop.first ? true : false}]
	) %}
{% endfor %}

{% if recipientsArray|length > 0 %}
	{# Populate field with values #}
	{% do form.setFieldSettings('fieldHandle', {
		options: recipientsArray
	}) %}
{% endif %}
```

However, when selecting a value from the dropdown and submitting the form, I get a 'Sending form notification' with following error, where [id:1] is the literal value in the option on the frontend, but should be converted to the proper email address after submit.

```
 Failed to send notification email: {"error":"Notification email parse error for “To: {field.dropdownLocationRecipients}”. Template error: “Address in mailbox given [id:1] does not comply with RFC 2822, 3.6.2.” /.../vendor/swiftmailer/swiftmailer/lib/classes/Swift/Mime/Headers/MailboxHeader.php:355"}. 
```

It seems that the internal array that is created (with the obscured email addresses) is not getting read when the email is send? The value ([id:1]) is passed through instead of being converted to the actual email address.

For a sanity check: a Recipients field with 'Dropdown' as its display type and with some options defined work just fine.

**Steps to reproduce**
1. Populate a recipients field with type dropdown using proper syntax in a template
2. Select an option on the frontend and submit the form

### Form settings
Please pick from the options below.

- Multi-page form: No
- Submission Method: Ajax
- Client-side Validation: Yes
- Custom Form Templates: No

**Additional info**
- Plugin version: 1.4.16
- Craft version: 3.7.13
- Multi-site: No

**Additional context**

The Recipients field is set as the "From" email address in the Email Notification set to this form.

**Comment by engram-design** [comment]
Does your Recipients field have any options? The only way I can replicate this is if the Recipients field has no options, and you populate it dynamically. If these are values that are dynamically generated, and don't exist on the field options when you create it, this will likely cause some issues. Refer to the [docs](https://verbb.io/craft-plugins/formie/docs/template-guides/rendering-fields#override-field-settings) and [discussion](https://github.com/verbb/formie/issues/415#issuecomment-815438161)

If however, you have a consistent set of options for the field (say 10 defined in the field) and you just want to filter them based on your entries query, that's fine. So long as all possible values exist in the field in the control panel.

For any dynamic behaviour, we typically recommend using a hidden field, and styling it like a dropdown (if you need it to be a dropdown). If you want the benefit of the field being hidden from the HTML source, you can set any fields' visibility to "Disabled" which doesn't render on the front-end. In your case, you could also swap it to a hidden Recipients field, but it's possible your use-case wants to allow users to select alternative emails in a dropdown (but maybe have a sane default based on your entry query).

**Comment by jimirobaer** [comment]
No, I dynamically populate the Recipients field from my templates; no options are provided from the editor. 

What is the best way to build a form via Formie where:

1. You can choose via a field on the front-end where the email should be send to (e.g. Sales, Marketing) ...
2. ... and that field should be built dynamically (from e.g. category, section).

I can't link the active (chosen) value from a dropdown to the 'From' email address in the editor. 

> For any dynamic behaviour, we typically recommend using a hidden field, and styling it like a dropdown (if you need it to be a dropdown). If you want the benefit of the field being hidden from the HTML source, you can set any fields' visibility to "Disabled" which doesn't render on the front-end. In your case, you could also swap it to a hidden Recipients field, but it's possible your use-case wants to allow users to select alternative emails in a dropdown (but maybe have a sane default based on your entry query).

Overriding the styling of a hidden field to appear on the page seems an odd solution. Could I use Javascript to write the chosen option of a dropdown (dynamically) to a hidden field after which that value is used to mail to?

**Comment by engram-design** [comment]
You can't dynamically populate a dropdown (or checkboxes/radio) field at this time. For the reasons outlined in [this discussion](https://github.com/verbb/formie/issues/415#issuecomment-815438161).

The reason for the hidden field is to prevent validation issues on the server. If you try to submit a value for a dropdown field that isn't within it's set options, then that's invalid. This helps with data consistency. Imagine if when viewing submissions, there's all sorts of values for a dropdown field, because they've been dynamically generated. This also means that technically, you could populate all sorts of "bad" or invalid values for a dropdown field. This defeats the purpose of a dropdown field, where you can only select from a defined set of options - not just **any** value.

So visually, the field can be shown as a select, but stored in a hidden field. But yes, that's another option - you could totally use JS to listen to the `change` event of a dropdown, and assign that to a hidden input. However, you'll want to exclude the dropdown field from the payload, as its dynamic value will be invalid when submitted.

**But** we may very well change this behaviour _just_ for the Recipients field, as this is arguably the desired behaviour (to be able to dynamically set the recipient). This is just a limitation of the dropdown/checkboxes/radio styling of this field. This is planned when we tackle [this](https://github.com/verbb/formie/issues/478)

Also note that I was suggesting the "Hidden" style for the recipients field, not an actual hidden field, but they are more or less the same.

**Comment by jimirobaer** [comment]
> So visually, the field can be shown as a select, but stored in a hidden field. But yes, that's another option - you could totally use JS to listen to the `change` event of a dropdown, and assign that to a hidden input. However, you'll want to exclude the dropdown field from the payload, as its dynamic value will be invalid when submitted.

How can I exclude a field from the payload? There's the option `Include in Email Notifications` available on Hidden Fields, but there's no such toggle on the Dropdown field. Because Formie now tries to validate the Dropdown field that I supplied with options dynamically, as in issue https://github.com/verbb/formie/issues/415.

Or do you mean I put a  select input field, separate from Formie on the page? If so, that would mean not being able to use the `craft.formie.renderForm()` function and that kind of defeats the purpose of using a form builder.

**Comment by engram-design** [comment]
So yep, I might've thrown you off by suggesting a Visibility = Hidden for a Recipients field, as you're right, you can't exclude that from an email. 

The best bet is a hidden field type in your form builder, but when you go to outputting it, use a custom template to style it like a dropdown, or whatever your needs. Because it's a hidden field, you can then exclude it form the email notification. Similar to that discussion in https://github.com/verbb/formie/issues/415#issuecomment-815438161

**Comment by jimirobaer** [comment]
> The best bet is a hidden field type in your form builder, but when you go to outputting it, use a custom template to style it like a dropdown, or whatever your needs.

That seems to work. It doesn't feel right having to do it, but better than nothing I suppose. 

However, because I am looping over one of my section in the hidden.html and outputting their email addresses, I now have 100+ email addressed out in the open as values on my select's option tags. 

I am listening to the onchange event of this input, and when changed, I copy the selected option's value (=> email address) to a hidden Recipient's field's value. 

By doing it this way, I risk 1) having the email address in the open text and 2) being able to overwrite the email address that the form uses as the recipient by simply updating the value of this field with my Developer Tools. This doesn't seem quite right. 

I could encode the addresses with something like base64, but that doesn't fixed the second problem I have with this.

**Comment by engram-design** [comment]
I'll see if it's possible to populate the recipients field, separately to the dropdown functionality (which is coming to 1.5.0), as I agree it's not ideal in that scenario of exposing emails.

**Comment by engram-design** [comment]
Added in [1.5.0](https://github.com/verbb/formie/releases/tag/1.5.0)

**Comment by jimirobaer** [comment]
I had some time to apply the dynamically population of recipients in a dropdown to see if I could get it to work. Previously I've overwritten the 'hidden' field as suggested by you above. This works but resulted in email addresses being exposed in the DOM.

I still seem to get the same error as in my initial post, where the 'value' part of the dynamically added option is not being converted to an email address,.

``` 
Failed to send notification email: {"error":"Notification email parse error for “To: {field.dealers}”. Template error: “Address in mailbox given [id:0] does not comply with RFC 2822, 3.6.2.” .../vendor/swiftmailer/swiftmailer/lib/classes/Swift/Mime/Headers/MailboxHeader.php:355"}.
 ```

I have set the field `dealers` as the Formie needs to sent email notifications to.

I am merging the entries from an ElementQuery using the following syntax:

```
{% set recipientsArray = recipientsArray|merge(
    [{"label": entry.title, "value": entry.eMail, "isDefault": loop.first ? true : false}]
) %}

{% if recipientsArray|length > 0 %}
    {% do form.setFieldSettings('dealers', {
        options: recipientsArray
    }) %}
{% endif %}
```

**Comment by engram-design** [comment]
So this poses a tricky situation in that using `populateFormValues()` or `setFieldSettings()` to mess around with recipient fields inherently exposes the email address(es) you're setting. This is because `populateFormValues()` essentially just set's the default value on an input/select. This is true even if the field is set to hidden, as you've mentioned.

However, upon further reflection that's a bug. Settings a hidden recipient field with `populateFormValues()` should set a base64 encoded string, not the un-encoded email. This is fixed for the next release. To get this fix early, change your `verbb/formie` requirement in `composer.json` to:

```
"require": {
  "verbb/formie": "dev-craft-3 as 1.5.8",
  "...": "..."
}
```

Then run `composer update`.

As such, if you're looking to dynamically set the recipient field value based off something arbitrary, you should use the hidden display type for the field. This would be opposed to using the checkboxes/radio/dropdown display type, modifying the available options on the fly, and setting the value - all of which is possible, but unnecessary (unless you actually want users to be able to pick from a list of different options that change for different pages where your forms are shown).

```
{% do form.setFieldSettings('recipientsDropdown', {
    options: [
        { label: 'Dynamic Title 1', value: 'dynamic+1@email.com' },
        { label: 'Dynamic Title 2', value: 'dynamic+2@email.com' },
    ],
}) %}
```

So the example above _should_ also work too, if you wanted to go the more exhaustive route using non-hidden display type layouts for the field. This is because Formie will save the options you're defining as a snapshot against the submission. My testing shows that works as of the latest `craft-3` branch above.

**Comment by engram-design** [comment]
Fixed in [1.5.9](https://github.com/verbb/formie/releases/tag/1.5.9)
